### PR TITLE
cilium, ipvlan: implement missing host connectivity

### PR DESCRIPTION
**Pull request by borkmann**
This change is [Reviewable](https://reviewable.io/reviews/cilium/cilium/6715)

### CODE REVIEW

File: bpf/init.sh [hunk 1]
```diff
@@ -413,9 +424,8 @@ else
 	fi
 fi
 
+# BPF host program attachment
 case "${MODE}" in
```
Comment (aanm) [inline bpf/init.sh]: remove this case as it it for all modes

Comment (borkmann) [reply bpf/init.sh]: Yep, will do once I have policy properly working.

**Comment by borkmann** [comment]
test-me-please

**Comment by borkmann** [comment]
test-me-please

**Comment by borkmann** [comment]
test-me-please

File: bpf/init.sh [hunk 2]
```diff
@@ -413,17 +424,15 @@ else
...
-		POLICY_MAP="cilium_policy_reserved_${ID_HOST}"
-		OPTS="-DFROM_HOST -DFIXED_SRC_SECCTX=${ID_HOST} -DSECLABEL=${ID_HOST} -DPOLICY_MAP=${POLICY_MAP}"
-		bpf_load $HOST_DEV1 "$OPTS" "egress" bpf_netdev.c bpf_host.o from-netdev $CALLS_MAP
-		;;
-esac
+# bpf_host.o requires to see an updated node_config.h which includes ENCAP_IFINDEX
+CALLS_MAP="cilium_calls_netdev_ns_${ID_HOST}"
+POLICY_MAP="cilium_policy_reserved_${ID_HOST}"
+if [ "$MODE" == "ipvlan" ]; then
+	OPTS="-DENABLE_EXTRA_HOST_DEV=1 -DFROM_HOST -DFIXED_SRC_SECCTX=${ID_HOST} -DSECLABEL=${ID_HOST} -DPOLICY_MAP=${POLICY_MAP}"
```
Comment (joestringer) [inline bpf/init.sh]: Minor tidy, could this instead be more like

```
OPTS="....default set"
if [ "$MODE" == "ipvlan" ]; then
  OPTS += "ipvlan-specific opts"
fi
```

**Review by joestringer — APPROVED** [review]
LGTM, one minor cosmetic nit.

Would be nice to move away from configuring stuff in `init.sh` just for future - we do stuff based on headerfile contents and anything in `init.sh` is code that the Golang daemon isn't aware of. But again this is more of a general ask rather than something specific to this PR.

File: bpf/init.sh (hunk 2 above)
Comment (borkmann) [reply bpf/init.sh]: Agree, much better, will do.

**Comment by borkmann** [comment]
test-me-please

**Comment by brb** [comment]
test-me-please

File: bpf/init.sh [hunk 3]
```diff
@@ -413,17 +424,14 @@ else
...
-		OPTS="-DFROM_HOST -DFIXED_SRC_SECCTX=${ID_HOST} -DSECLABEL=${ID_HOST} -DPOLICY_MAP=${POLICY_MAP}"
-		bpf_load $HOST_DEV1 "$OPTS" "egress" bpf_netdev.c bpf_host.o from-netdev $CALLS_MAP
-		;;
-esac
+# bpf_host.o requires to see an updated node_config.h which includes ENCAP_IFINDEX
+CALLS_MAP="cilium_calls_netdev_ns_${ID_HOST}"
+POLICY_MAP="cilium_policy_reserved_${ID_HOST}"
+OPTS="-DFROM_HOST -DFIXED_SRC_SECCTX=${ID_HOST} -DSECLABEL=${ID_HOST} -DPOLICY_MAP=${POLICY_MAP}"
+if [ "$MODE" == "ipvlan" ]; then
+	OPTS+="-DENABLE_EXTRA_HOST_DEV"
```
Comment (brb) [inline bpf/init.sh]: A separating space is missing, i.e. `OPTS+=" -DENABLE_EXTRA_HOST_DEV"`. Otherwise, bpf_netdev.c won't compile.

**Review by brb — CHANGES_REQUESTED** [review]

**Comment by brb** [comment]
I've tested this with the ipvlan master dev set to a dummy netdev. ICMP and TCP works from host to container, and from container I'm able to ping `cilium_host` IPv4.

~However, when I set the ipvlan master dev to a real one, I'm not able to ping the container from the host.~ UPDATE: it works!

**Comment by borkmann** [comment]
test-me-please

**Review by brb — APPROVED** [review]

**Comment by borkmann** [comment]
test-me-please

**Comment by borkmann** [comment]
test-me-please

**Comment by borkmann** [comment]
test-me-please

**Comment by borkmann** [comment]
Hit #6728

**Comment by borkmann** [comment]
Prior test also hit #6727

**Comment by borkmann** [comment]
test-me-please

**Comment by borkmann** [comment]
#6727 again

**Comment by borkmann** [comment]
test-me-please

**Comment by borkmann** [comment]
CI green

**Pull request merged**